src/ray/object_manager/notification/object_store_notification_manager.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_OBJECT_STORE_NOTIFICATION_MANAGER_H
+#define RAY_OBJECT_STORE_NOTIFICATION_MANAGER_H
+
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include <boost/asio.hpp>
+
+#include "ray/common/id.h"
+#include "ray/common/status.h"
+#include "ray/object_manager/format/object_manager_generated.h"
+
+namespace ray {
+
+/// \class ObjectStoreNotificationManager
+///
+/// Encapsulates notification handling from the object store.
+class ObjectStoreNotificationManager {
+ public:
+  ObjectStoreNotificationManager(boost::asio::io_service &io_service)
+      : main_service_(&io_service), num_adds_processed_(0), num_removes_processed_(0) {}
+  virtual ~ObjectStoreNotificationManager() {}
+  /// Subscribe to notifications of objects added to local store.
+  /// Upon subscribing, the callback will be invoked for all objects that
+  /// already exist in the local store
+  ///
+  /// \param callback A callback expecting an ObjectID.
+  void SubscribeObjAdded(
+      std::function<void(const object_manager::protocol::ObjectInfoT &)> callback) {
+    add_handlers_.push_back(std::move(callback));
+  }
+
+  /// Subscribe to notifications of objects deleted from local store.
+  ///
+  /// \param callback A callback expecting an ObjectID.
+  void SubscribeObjDeleted(std::function<void(const ray::ObjectID &)> callback) {
+    rem_handlers_.push_back(std::move(callback));
+  }
+
+  /// Support for rebroadcasting object add/rem events.
+  void ProcessStoreAdd(const object_manager::protocol::ObjectInfoT &object_info) {
+    // TODO(suquark): Use strand in boost asio to enforce sequential execution.
+    for (auto &handler : add_handlers_) {
+      main_service_->post([handler, object_info]() { handler(object_info); });
+    }
```

Comment:
We could post the entire loop over the handlers as one callback. Would that work?

Reply:
yes, that should work. actually I just notice that we have (potential) concurrency issues here. the object manager could add new handlers while the plasma store thread is looping over the handlers. This will not actually happen here because we only have two fixed `add_handlers_ `, both created at the beginning.

Maybe a better choice is to enforce handlers passing-in during constructing?